### PR TITLE
Loki: Don't request stats with invalid query

### LINKED ISSUE #65276: Loki Stats: Do not call stats endpoint if the query is invalid

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
fixes a bug requesting the stats endpoint even if your query is invalid

**Why do we need this feature?**
...

**Who is this feature for?**
...

**Which issue(s) does this PR fix?**:
Fixes #65276

**Special notes for your reviewer:**

### CODE REVIEW

**Review by ivanahuckova — COMMENTED** [review]
nice! I am wondering if we could maybe do 1 more optimisation. In `shouldUpdateStats` in `stats.ts` we are comparing `query !== prevQuery`, but we care only about stream selectors, so what do you think about updating this to check if stream selector has changed? 🤔

**Review by matyax — COMMENTED** [review]
Can we have some tests to assert that these changes produce the expected output?

imagen

**Review by ivanahuckova — COMMENTED** [review]
I run the code and it still requests stats for invalid query 🤔
image

**Comment by gwdawson** [comment]
> I run the code and it still requests stats for invalid query 🤔

hmm, it seems like `{label="value",}` is an edge case that isn't caught by `isValidQuery()`.
do other invalid queries still send for you? or just this one?

File: public/app/plugins/datasource/loki/datasource.ts
Comment (svennergr) [inline public/app/plugins/datasource/loki/datasource.ts:429]: This might be a nit, but I think there's a semantic difference between `null` and `undefined`. I think in this case `null` would be rather the value we should return, as we deliberately chose to return "defined nothing".

**Comment by svennergr** [comment]
> > I run the code and it still requests stats for invalid query 🤔
> 
> hmm, it seems like `{label="value",}` is an edge case that isn't caught by `isValidQuery()`. do other invalid queries still send for you? or just this one?

I think that's a bug here: That line is not necessary https://github.com/grafana/lezer-logql/blob/main/src/logql.grammar#L25

**Comment by gwdawson** [comment]
currently blocked by #65436 failing tests.

**Comment by matyax** [comment]
https://github.com/grafana/grafana/pull/66289 will update Lezer and the autocomplete to unblock.

**Comment by gwdawson** [comment]
@matyax thanks for the fix, this pr works as expected now

**Review by ivanahuckova — APPROVED** [review]
LGTM 🚀

**Pull request merged**
